packages/core/src/typeschema/crawler.ts

```diff
@@ -103,27 +107,78 @@ class ResourceCrawler {
 }
 
 export function getNestedProperty(
-  value: TypedValue,
+  value: TypedValueWithExpression | undefined,
   key: string,
-  options?: { profileUrl?: string }
+  options: { profileUrl?: string; withExpression: true }
+): (TypedValueWithExpression | TypedValueWithExpression[])[];
+export function getNestedProperty(
+  value: TypedValue | undefined,
+  key: string,
+  options?: { profileUrl?: string; withExpression?: false }
+): (TypedValue | TypedValue[] | undefined)[];
+export function getNestedProperty(
+  value: TypedValue | undefined,
+  key: string,
+  options?: { profileUrl?: string; withExpression?: boolean }
 ): (TypedValue | TypedValue[] | undefined)[] {
+  if (value === undefined) {
+    return [undefined];
+  }
+
   if (key === '$this') {
     return [value];
   }
+
+  const propertyGetter = options?.withExpression ? getTypedPropertyValueWithExpression : getTypedPropertyValue;
+
   const [firstProp, ...nestedProps] = key.split('.');
-  let propertyValues = [getTypedPropertyValue(value, firstProp, options)];
+  let propertyValues = [propertyGetter(value, firstProp, options)];
   for (const prop of nestedProps) {
     const next = [];
     for (const current of propertyValues) {
       if (Array.isArray(current)) {
         for (const element of current) {
-          next.push(getTypedPropertyValue(element, prop, options));
+          next.push(propertyGetter(element, prop, options));
         }
-      } else if (current !== undefined) {
-        next.push(getTypedPropertyValue(current, prop, options));
+      } else if (options?.withExpression && current && current.value !== undefined) {
+        next.push(propertyGetter(current, prop, options));
+      } else if (!options?.withExpression && current !== undefined) {
+        next.push(propertyGetter(current, prop, options));
       }
     }
     propertyValues = next;
   }
   return propertyValues;
 }
+
+function getTypedPropertyValueWithExpression(
+  input: TypedValue | TypedValueWithExpression,
+  path: string,
+  options?: GetTypedPropertyValueOptions
+): TypedValueWithExpression[] | TypedValueWithExpression {
+  const parentExpression = (input as TypedValueWithExpression).expression;
+  return withExpression(getTypedPropertyValue(input, path, options), parentExpression, path);
+}
+
+export type TypedValueWithExpression = TypedValue & { expression: string };
```

Comment:
I would suggest "path" over "expression" for this concept to more accurately describe what's being tracked, given that these are explicitly paths into the resource

Reply:
Summarizing an offline discussion: since `path` would collide with the same-named property on React component inputs, `expression` is the next most specific option, so going with that for now.